// Copyright 2015 gRPC authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

syntax = "proto3";

option java_multiple_files = true;
//option java_package = "io.grpc.examples.helloworld";
//option java_outer_classname = "HelloWorldProto";
//option objc_class_prefix = "HLW";


package BankService;

// The bank service definition.
service BankService {
    // Method for communication between Customer and Branch
    rpc MsgDelivery(CustomerRequest) returns (ResponseMessage);

    // Method to query the balance of a Branch
    rpc Query(BalanceQuery) returns (ResponseMessage);

    // Method to withdraw from a Branch
    rpc Withdraw(CustomerRequest) returns (ResponseMessage);

    // Method to deposit to a Branch
    rpc Deposit(CustomerRequest) returns (ResponseMessage);

    // Method for Branch to handle requests from fellow branches
    rpc Propagate_Withdraw(BranchRequest) returns (ResponseMessage);

    // Method for Branch to handle requests from fellow branches
    rpc Propagate_Deposit(BranchRequest) returns (ResponseMessage);
}

message CustomerRequest {
    int32 customer_id = 1;
    string event = 2;
    int32 amount = 3;
}

message BranchRequest {
    int32 branch_id = 1;
    int32 amount = 2;
}

message ResponseMessage {
    string message = 1;
    bool success = 2;
}

message BalanceQuery {
    int32 branch_id = 1;
}